### PR TITLE
[goecharger] Add API V2 support

### PR DESCRIPTION
**Pull request by ranzen84**
Add API v2 support for the goecharger-binding (the new hardware revision 3 of the go-eCharger is supporting API v2):

-new thing configuration parameter "apiVersion" (default will be 1, as binding did before)
-add filter for reading only needed parameters to improve performance (API v2)
-add some parameters/channels when using API v1
-some logging changes
-handle InterruptedException explicitly

### CODE REVIEW

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:259]: See previous comment about thing state and logger level

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:300]: See previous comment about lower trace level.

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:304]: See previous comment about thing state and logger level

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:307]: See previous comment about thing state, logger level and handling exceptions.

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:331]: Trace here too

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:312]: Please use consistent naming of the charger, the readme uses Go-eCharger. Several other pleases too.

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml.bak
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml.bak:1]: Possibly accidently added BAK file.

**Review by lsiepel — CHANGES_REQUESTED** [review]
Thanks for adding this addition to the binding. I left some comments, most of them are minor issue's. 
Please don't forget to sign the commit and someone more experienced with reviewing will look into this.

File: bundles/org.openhab.binding.goecharger/README.md
Comment (ranzen84) [reply bundles/org.openhab.binding.goecharger/README.md]: API needs to be activated via the Go-eCharger app. I think this should be done conscious

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (ranzen84) [reply bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:259]: This is the situation if you try to set a key that is not allowed, the thing state will remain unchanged (ONLINE) then

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:314]: This method has some overloads, so that you can add a statusdetail to let the user know why it is offline.

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:318]: also add the ThingStatusDetail

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:319]: I still think this needs some attention.

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:309]: I still think this needs some attention. Please also set the ThingStatusDetail

**Review by lsiepel — CHANGES_REQUESTED** [review]
Just finished my second pass. Looks like most of the comments are fixed. Don't forget to sign-off your commits. All three commits seem not correctly signed. Did you also fix the SAT errors?

**Comment by ranzen84** [comment]
I'm not sure what I did wrong with the sign-off? If I have a look at the commit, it is there.

**Comment by lsiepel** [comment]
> I'm not sure what I did wrong with the sign-off? If I have a look at the commit, it is there.

If you click details next to the DCO line you get this link: https://github.com/openhab/openhab-addons/pull/12400/checks?check_run_id=5418393079 It shows the DCO details with a summary on the lower part of the page:

`Commit sha: [62afc13](https://github.com/openhab/openhab-addons/pull/12400/commits/62afc13c347c0e66823540c6a21acf6265cbfae0), Author: ranzen84, Committer: ranzen84; Expected "ranzen84 [reinhardplaim@gmail.com]", but got "Reinhard Plaim [reinhardplaim@gmail.com]".`

So i guess a mismatch between your github details and you certificate details used with gpg signing the commits. I'm not sure how to re-sign the previous commits. A good starting point: https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/GoEChargerConfiguration.java
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/GoEChargerConfiguration.java:29]: If I understand this correctly, this is the default API version? For hardware revisions supporting v2, is v1 also supported? If this is the case, perhaps it would make more sense to use v1 as default for better compatibility, since I would assume older hardware revisions only supports v1?

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerV2Handler.java:346]: Second parameter is redundant.

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:314]: Please add i18n string for localization.

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:26]: Is this really needed as a channel? Usually firmware version is provided as a property.

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:43]: You can use attributes `min` and `max` is you want the configuration validated so only values 1 and 2 will be allowed.

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:199]: Consider adding category **Temperature**.

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:43]: Consider if this parameter could be optional so you could default to 1. This would provide backwards compatibility for existing users.

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerBaseHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerBaseHandler.java:54]: Can this class be `abstract`, since you probably only want to allow instances of subclasses?

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:318]: Consider if `e.toString()` would make sense to provide as third parameter.

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:370]: For `InterruptedException`, please call `Thread.currentThread().interrupt()`. See #12365.

**Review by jlaur — CHANGES_REQUESTED** [review]
Please see comments.

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (ranzen84) [reply bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:26]: Sorry, the difference is not clear for me. Can you give me a hint or link?

Comment (lsiepel) [reply bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:26]: https://www.openhab.org/docs/developer/bindings/#properties

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerBaseHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerBaseHandler.java:143]: I think this has been commented before. Please handle the excpetions on their own and atleast (for interruptexception) call Thread.currentThread().interrupt();

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [reply bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:26]: Sorry, this channel also existed before, so it's okay to leave it for compatibility.

**Comment by jlaur** [comment]
Please see #12391 also as these two PR's are conflicting. The other one seems to add another temperature channel for API v1, which for this PR is added only for API v2. Maybe something could be picked for this PR?

**Comment by ranzen84** [comment]
I think we do not need the 2 temperatures for API v1, cause this is only available if you have hardware revision 3. This means you can activate API v2, where the 2 temps are available.
If you have an older hardware revision, only one temp is available.

https://github.com/goecharger/go-eCharger-API-v1
https://github.com/goecharger/go-eCharger-API-v2/blob/main/introduction-de.md

**Comment by ranzen84** [comment]
I can add the other parameter for API v1 of https://github.com/openhab/openhab-addons/pull/12391 
"maxCurrTemp" if we need it. This parameter is not available for API v2

**Comment by jlaur** [comment]
> I can add the other parameter for API v1 of #12391
> "maxCurrTemp" if we need it. This parameter is not available for API v2

It's up to you if you want to do it in context of this PR. One last thing though: Since PR is not linked to an issue, can you write a short description for the PR? The title doesn't seem to cover the changes as new channels have also been introduced, for example.

After this PR is merged, please also create a new PR towards openhab/openhab-distro with a compatibility comment - like this: openhab/openhab-distro#1369. When channel types are renamed (i.e. old channel type id doesn't exist anymore), things created from UI will break.

**Comment by jlaur** [comment]
> I can add the other parameter for API v1 of #12391
> "maxCurrTemp" if we need it. This parameter is not available for API v2

It's up to you if you want to do it in context of this PR. One last thing though: Since PR is not linked to an issue, can you write a short description for the PR? The title doesn't seem to cover the changes as new channels have also been introduced, for example.

After this PR is merged, please also create a new PR towards openhab/openhab-distro with a compatibility comment - like this: openhab/openhab-distro#1369. When channel types are renamed (i.e. old channel type id doesn't exist anymore), things created from UI will break.

**Comment by ranzen84** [comment]
added parameters from other PR, description for this PR will follow

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:322]: Just curious: Does this mean that it didn't work previously, or does it mean that both GET and POST will work? In other words: What is the reason for this change?

Comment (ranzen84) [reply bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:322]: I took this change from the other PR, I will try to test it with my charger with hardware revision 3 and API v1

**Comment by ranzen84** [comment]
Cause of the change of the channel-type ids: I was not aware of the impact. I will restore to old/existing channel-type ids if possible

File: bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java
Comment (ranzen84) [reply bundles/org.openhab.binding.goecharger/src/main/java/org/openhab/binding/goecharger/internal/handler/GoEChargerHandler.java:322]: working both with GET and POST

**Comment by ranzen84** [comment]
from my point of view, everything is done for this PR now, or did I missed something?

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:217]: Is this rename by purpose or left by mistake? Now that the other channel type id's are renamed back.

File: bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml
Comment (jlaur) [inline bundles/org.openhab.binding.goecharger/src/main/resources/OH-INF/thing/thing-types.xml:44]: Documented as optional since it should default to 1. Please check if it should be required here.

**Comment by navimike** [comment]
Hi guys,
Please can we add also support for doing the switch of the phases (relevant for V3 chargers)? The API parameter is "psm". I did successful tests using the API V2 commands below. I'm currently doing a manual HTTP call in my javascripts, but I would prefer to have it "clean" in the binding. 
http://192.168.0.22/api/set?psm=2 -> 3 phases
http://192.168.0.22/api/set?psm=1 -> 1 phase

Best regards,
Michael

**Comment by ranzen84** [comment]
psm is included already in this PR, if you want to set it, you can take 3 or 1 (instead of 2 and 1)

**Comment by navimike** [comment]
> psm is included already in this PR, if you want to set it, you can take 3 or 1 (instead of 2 and 1) I will add a second example for a rule

Great to hear! Thank you!
3 and 1 is also more meaningful to me. I'm also a developer (unfortunately not much experience on open source). Please let me know, if I can you support in any way.

Cheers!

**Review by jlaur — APPROVED** [review]
Thanks for these improvements!

**Comment by chilobo** [comment]
Thanks for working on this. I never got my version merged. (Which works locally with API 1 and hardware version 3.
And it doesn't support API v.2 which this one does.)

I will try to integrate your binding in my openhab-setting on the weekend and will give feedback. Unfortunatly I only have a 1p-car, so I cannot test switching to 3p.

**Comment by chilobo** [comment]
I tried to use your binding today upgrading my openhab on an Raspi to 
Version:     3.3.0-SNAPSHOT (#2797)
bundle:list shows:
3.3.0.202203110340     â openHAB Add-ons :: Bundles :: Go-eCharger Binding
In the GUI I added the binding, the new one is shown, documentation states:

`Thing Configuration
The thing has two configuration parameters:

Parameter	Description	Required
ip	The IP-address of your Go-eCharger	yes
apiVersion	The API version to use (1=default or 2)	no
refreshInterval	Interval to read data, default 5 (in seconds)	no
`
Btw: Should it be: **three** configurations parameters?

When I try to add a thing (by Administration - Einstellungen - Things - +) and select GO-echarger binding it says 
'Add manually 
No thing types can be added with this binding.'

**Comment by lsiepel** [comment]
As this PR is merged, could you raise an issue or discuss this in the openHAB community forums?

**Pull request merged**
